nixos/doc: improve install instructions

davidak opened the pull request:
###### Description of changes

- Update download URLs
- Replace "USB stick"/"USB Drive" with "USB flash drive" as that seem more correct
  https://en.wikipedia.org/wiki/USB_flash_drive
  https://elementary.io/docs/installation#choose-operating-system
- Don't mention CD as easiest option anymore,
  as all modern systems should be able to boot from USB,
  but many don't have a CD drive. Burning CDs is also usually wasteful as you
  can't burn them again.
- Remove link to NixOS Wiki (Making_the_installation_media) as it is not needed
- Add Etcher and USBImager as graphical tools to create install drive
- Make dd command consistent and use block size of 4 MB for faster flashing
- More consistent text
- Add instructions for "Booting from the install medium"
  Inspired by https://github.com/elementary/website/blob/9a91b0f4956b059db02122fbc738c4c067e033a7/docs/installation.md#booting-from-the-install-drive-booting-from-the-installation-medium-clear-float-2
- Add instructions for "Graphical Installation"
- Restructure headings and anchors for "Manual Installation"
- Adding legacy anchors for "Manual Installation" to not break links

---

I have tested booting from install drive created with Etcher, USBImager and dd on linux. It would be great if someone can test the other options (macOS and Windows).

I hope to get this into the release.

cc @dasJ 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

davidak committed:
nixos/doc: improve install instructions

- Update download URLs
- Replace "USB stick"/"USB Drive" with "USB flash drive" as that seem more correct

  https://en.wikipedia.org/wiki/USB_flash_drive
  https://elementary.io/docs/installation#choose-operating-system

- Don't mention CD as easiest option anymore,
  as all modern systems should be able to boot from USB,
  but many don't have a CD drive. Burning CDs is also usually wasteful as you
  can't burn them again.
- Remove link to NixOS Wiki (Making_the_installation_media) as it is not needed
- Add Etcher and USBImager as graphical tools to create install drive
- Make dd command consistent and use block size of 4 MB for faster flashing
- More consistent text
- Add instructions for "Booting from the install medium"

  Inspired by https://github.com/elementary/website/blob/9a91b0f4956b059db02122fbc738c4c067e033a7/docs/installation.md#booting-from-the-install-drive-booting-from-the-installation-medium-clear-float-2

- Add instructions for "Graphical Installation"
- Restructure headings and anchors for "Manual Installation"
- Adding legacy anchors for "Manual Installation" to not break links

Co-authored-by: j-k <dev@j-k.io>
Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
Co-authored-by: Robert Schütz <github@dotlambda.de>
Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>
Co-authored-by: Thiago Kenji Okada <thiagokokada@gmail.com>

jtojnar committed:
nixos/doc/manual/md-to-db.sh: Add support for <kbd> element

jtojnar commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -33,7 +60,7 @@ To install over a serial port connect with `115200n8` (e.g.
 entries, select the serial console boot entry.
 
-### Networking in the installer {#sec-installation-booting-networking}
+### Networking in the installer {#sec-installation-manual-networking}
 
```
Maybe also preserve the old anchor to keep external links working:

```suggestion
### Networking in the installer {#sec-installation-manual-networking}
[]{#sec-installation-booting-networking} <!-- legacy anchor -->

```

davidak replied:
I thought about that. Thanks for showing how to do it.

samueldr commented on nixos/doc/manual/installation/installing-usb.section.md:
```diff
@@ -1,31 +1,49 @@
+with write-only (wo) interface for your system. Start the program,
+select the image, select the USB flash drive and click "Write".
+
+## Creating bootable USB flash drive from Terminal on Linux {#sec-booting-from-usb-linux}
```
```suggestion
## Creating bootable USB flash drive from the Terminal on Linux {#sec-booting-from-usb-linux}
```

or *a* terminal.

(See also the macOS section)

samueldr commented on nixos/doc/manual/installation/installing-usb.section.md:
```diff
@@ -1,31 +1,49 @@
-$ diskutil unmountDisk diskN
-Unmount of all volumes on diskN was successful
-$ sudo dd if=nix.iso of=/dev/rdiskN bs=1M
+sudo dd if=<path-to-image> of=/dev/sdX bs=4M conv=fsync oflag=direct status=progress
```
We might not want to add the `bs`, `conv`, `oflag` and `status` parameters as they might not be present in all `dd` flavours used by distros.

Also, other commenters: check the manpage for `dd` and cite from it if you have any concerns about the params. A lot of "common knowledge" around `dd` is cargo-culted and not exactly right.

davidak replied:
I actually just copied the command from Arch Wiki. The progress information seem useful to me, but i can remove that again for compatibility.

samueldr commented on nixos/doc/manual/installation/installing-usb.section.md:
````diff
@@ -1,31 +1,49 @@
+
+```ShellSession
+diskutil unmountDisk diskX
+sudo dd if=<path-to-image> of=/dev/rdiskX bs=1m
````
is `1m` right? the case may be important with this `dd`.

And shouldn't we recommend the same block size on all platforms?

davidak replied:
`1m` is correct for macOS dd. dd from gnu uses `1M`

i did not change that command, so it should work well. not sure if we can safely remove the `bs`

bronson replied:
dd defaults to bs=512, which is really slow:

```sh
# transfer 1GB with bs=4m
% time dd if=/dev/zero of=/dev/null bs=4194304 count=256
256+0 records in
256+0 records out
1073741824 bytes transferred in 0.043489 secs (24689975261 bytes/sec)
dd if=/dev/zero of=/dev/null bs=4194304 count=256
0.00s user 0.04s system 96% cpu 0.046 total

# transfer 1GB with bs=512 (the default)
% time dd if=/dev/zero of=/dev/null bs=512 count=2097152
2097152+0 records in
2097152+0 records out
1073741824 bytes transferred in 16.644957 secs (64508538 bytes/sec)
dd if=/dev/zero of=/dev/null bs=512 count=2097152 
0.91s user 13.67s system 87% cpu 16.648 total
```

So on this hypothetical zero-latency disk, bs=4M produces around a 350X speedup. In the real world I've seen 10X faster.

If the issue is Gnu vs. BSD suffixes, can you use `bs=4194304`?

davidak replied:
Thanks for the explanation! Then we should use 4m for both.

We have different commands for GNU and BSD, so we can use the short version, they only have to be correct.

samueldr commented on nixos/doc/manual/installation/installing-usb.section.md:
```diff
@@ -1,31 +1,49 @@
+
+## Creating bootable USB flash drive from Terminal on macOS {#sec-booting-from-usb-macos}
+
+Use the `dd` utility to write the image to the USB flash drive. Unmount the disk before.
```
That last sentence feels *off*. It should anyway probably be written before the instruction to use `dd`. Something like *make sure the drive is unmounted.*, or any other preferred formulation.

This is probably needed in the Linux instructions too.

davidak replied:
to test if unmounting is needed, i have created 3 fat partitions on a usb drive, mounted them in my file browser (elementary Files, which use GVFS), created some folders and copied some files, then used dd to write the image, the partitions where still mounted and i was able to copy files, but unmounting takes very long... then i see only one partition (the first), i can mount it again, but all files copied after the dd are messed up

![Screenshot from 2022-05-29 18-56-43](https://user-images.githubusercontent.com/91113/170882256-b5acd2f3-7f2d-4e52-87d9-0bf92bf024ff.png)

GNOME Files still shows the old partitions before the dd

when disconnecting the drive and plugging it back in, it looks correct. so the dd should work without unmounting, but your programs will be confused about the mounted partitions and you probably loose data copied there into the void

i add instructions to unmount

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
-The NixOS manual is available by running `nixos-help`.
+     ::: {.tip}
+     If you don't have a favorite desktop and don't know which to choose, you
+     can stick to either GNOME or Plasma. They are quiet different from their
```
```suggestion
     can stick to either GNOME or Plasma. They are quite different from their
```

Quiet is silent.

```suggestion
     can stick to either GNOME or Plasma. They are quite different in their
```

I don't think "different from their design" is proper english, but I'm also an english-as-a-second-language person...

davidak replied:
right. "in their design" sounds more correct

i just use "They have a quite different design" now

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
+6.   You have the option to allow unfree software in the next screen.
+
+7.   The easiest setting in the "Partitioning" screen is "Erase disk", which will
+     delete all data from the selected disk and installs the system on it.
```
Either

```suggestion
     delete all data from the selected disk, and installs the system on it.
```

or

```suggestion
     delete all data from the selected disk and install the system on it.
```

It depends though on how the sentence is split I guess? I assume the difference is between "delete and install" or "delete, then installs".

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
-differences are mentioned in the steps that follow.
+To begin the installation, you have to boot your computer from the install drive.
+
+1.   Assuming your computer is still running, plug in the install drive
+and restart your computer.
```
So from power-off I don't need to plug in the install drive?

(Not sure how to rephrase this, but it seems like there's a disconnect between the sentences.)

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
+1.   Assuming your computer is still running, plug in the install drive
+and restart your computer.
+
+2.   Most computers will allow to select a drive to temporary boot from by
+pressing a special key, usually <kbd>F12</kbd>, but sometimes <kbd>Esc</kbd>
+or another function key. Refer to the screen or your computer's documentation.
```
This is not a discrete step, merge down with the next step.

davidak replied:
i think finding the right key to press and actually pressing it can be seen as two different tasks. it's already a lot of text for these steps. merging them would be a lot of text

samueldr replied:
Would a `::: {.note}` callout or similar construct help here?

The way I see it, the step is to boot the installation media, a *tip* or *note* can add additional details.

davidak replied:
very good suggestion! i'll add that

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
+3.   Press <kbd>F12</kbd> (or the appropriate key) and select the install drive,
+usually "USB-HDD" or something containing the word "USB", but the wording may vary.
+If your computer support BIOS and UEFI boot, choose the UEFI option.
+If you choose the incorrect drive, your computer will likely continue to boot
+as normal. Just restart your computer and pick a different drive in that menu.
```
*Just*?

```suggestion
If you choose the incorrect drive, your computer will likely continue to boot
as normal. In that case restart your computer and pick a different drive in that menu.
```

*Just* is almost always a poor choice when connecting sentences or ideas in docs, IMO. Here it could be understood as passing a judgement that "it's simple, *just* pick the right thing". A more neutral formulation feels better IMO.

davidak replied:
yes, it can be very frustrating when you do that for the first time and it does not work as you want

samueldr commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,127 @@
+(or press <kbd>Enter</kbd> to speed up).
+
+5.   The graphical images will start the desktop environment and the graphical
+installer, which can take some time. The minimal images will just have
+a command line.
```
```suggestion
installer, which can take some time. The minimal images will boot to
a command line.
```

The usage of *just* here is fine, but I'm proposing this alternate formulation that, I think, better describes what happens.

davidak replied:
good suggestion!

davidak commented:
@samueldr thank you for the review!

samueldr commented on nixos/doc/manual/installation/installing-usb.section.md:
````diff
@@ -1,31 +1,54 @@
+the `dd` utility to write the image to the USB flash drive:
+
+```ShellSession
+sudo bash -c 'for p in /dev/sdX*; do umount $p; done'
````
```suggestion
sudo umount /dev/sdX*
```

Rationale:

```
~ $ mount | grep /dev/sdb
/dev/sdb2 on /mnt type ext4 (rw,relatime)
/dev/sdb1 on /mnt2 type ext4 (rw,relatime)

~ $ sudo umount /dev/sdb*
umount: /dev/sdb: not mounted.
umount: /dev/sdb3: not mounted.

~ $ mount | grep /dev/sdb

~ $
```

dasJ commented:
I find it unlikely that this will get into the release in time since its a fairly new PR touching a lot of docs and I assume more people want to review it (release schedule is for tomorrow).
That being said, I see no reason we shouldn't backport this at a later point. I don't assume a lot more people will do new installations after the 22.05 release so there's no need to hurry and more time for thorough reviews of this critical piece of docs (one of the first items a new user would touch)

dasJ commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,130 @@
 # Installing NixOS {#sec-installation}
 
-## Booting the system {#sec-installation-booting}
+## Booting from the Install drive {#sec-installation-booting}
```
```suggestion
## Booting from the install medium {#sec-installation-booting}
```

dasJ commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,130 @@
+     wording may vary. If your computer support BIOS and UEFI boot, choose the
+     UEFI option. If you choose the incorrect drive, your computer will likely
+     continue to boot as normal. In that case restart your computer and pick
+     a different drive in that menu.
```
```suggestion
     a different drive in the menu.
```

dasJ commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,130 @@
+     a different drive in that menu.
+
+4.   Shortly after selecting the appropriate boot drive, you should be
+     presented with a menu with different Installer options. Leave the default
```
```suggestion
     presented with a menu with different installer options. Leave the default
```

dasJ commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,130 @@
+     presented with a menu with different Installer options. Leave the default
+     and wait (or press <kbd>Enter</kbd> to speed up).
+
+5.   The graphical images will start the desktop environment and the graphical
```
```suggestion
5.   The graphical images will start their corresponding desktop environment and the graphical
```

davidak commented:
> I don't assume a lot more people will do new installations after the 22.05 release
>this critical piece of docs (one of the first items a new user would touch)

i think that is a good reason to merge this big improvement before the release. two topics are not documented at all right now

it might not be perfect, but i think the experience with this change would be better than without it

do you see specific parts that can be controversial? (i will add system requirements in a separate PR)

06kellyjac commented on nixos/doc/manual/from_md/installation/installing-usb.section.xml:
```diff
@@ -1,35 +1,82 @@
-Unmount of all volumes on diskN was successful
-$ sudo dd if=nix.iso of=/dev/rdiskN bs=1M
+sudo umount /dev/sdX*
+sudo dd if=&lt;path-to-image&gt; of=/dev/sdX bs=4M
```
It would be good to add the progress `status=progress`

https://wiki.archlinux.org/title/USB_flash_installation_medium#In_GNU/Linux

A lot of guides just suggest to use `cat` these days but there's some nuance to using `sudo` + `>` so I think `dd` is more straightforward

davidak replied:
We  discussed the command before: https://github.com/NixOS/nixpkgs/pull/175128#discussion_r884184464

@samueldr what do you think about `progress`? would it be safe to add

SuperSandro2000 replied:
```suggestion
sudo dd if=&lt;path-to-image&gt; of=/dev/sdX bs=4M status=progress
```

samueldr replied:
It's likely fine to add, yes.

As I tried to explain in the initial (now orphaned) comment, there is a lot of cargo-culting, and selecting options is *hard*. Especially when there are multiple divergent `dd` implementations.

Though it's unlikely to cause issues to add those. Hopefully users using `dd` know what to do when commonly used options don't work.

SuperSandro2000 commented on nixos/doc/manual/installation/obtaining.chapter.md:
```diff
@@ -1,24 +1,21 @@
+[](#sec-booting-from-usb) to create a bootable USB flash drive.
+
+If you have a very old system that can't boot from USB, you can burn the image
+to an empty CD. NixOS might not work very well on such old systems.
```
```suggestion
to an empty CD. NixOS might not work very well on such systems.
```

davidak replied:
do you mean it should be deduped?

Mic92 replied:
I think he meant dedup here.

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
-to continue on the terminal, you can use `loadkeys` to switch to your
-preferred keyboard layout. (We even provide neo2 via `loadkeys de
-neo`!)
+You can use `loadkeys` to switch to your preferred keyboard layout.
```
Is there a difference to setxkbmap?

samueldr replied:
Yes.

`loadkeys` works on the kernel fbcon. `setxkbmap` works on X11.

davidak replied:
i have no idea. i have not changed the command, so maybe discuss that in a new issue

it does work, so i don't know if we should change it

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+
+1.   Plug in the install drive. Then turn on or restart your computer.
+
+2.   Press <kbd>F12</kbd> (or the appropriate key) to open the boot menu.
```
```suggestion
2.   Open the boot menu by pressing the appropriate key that's shown on your monitor on early boot. If no key is displayed consult your motherboard or system handbook.
```
That key differs widely across many motherboard and laptop manufactures. It could also be F1, F9, F10, Enter or Del.

Probably needs to be deduped with the warning.

davidak replied:
i had there first to look it up in the computers documentation, but i think it's easier to look it up online. most people probably don't find these documentations

i use your suggestion for the first sentence

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+
+2.   Press <kbd>F12</kbd> (or the appropriate key) to open the boot menu.
+     Select the USB flash drive (the option usually contains the word "USB").
+     If you choose the incorrect drive, your computer will likely continue to
+     boot as normal. In that case restart your compute and pick a different drive.
```
```suggestion
     If you choose the incorrect drive, your computer will either continue to
     boot as normal or abort with no OS found. In both cases restart your compute and pick a different drive.
```

samueldr replied:
In addition to suggestions, *your comput* → *your compute**r***.

davidak replied:
i think that is a rare case and does not add much useful information. the "likely" says that other things can happen, but they are not relevant to mention here

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+
+     ::: {.note}
+     The key to open the boot menu is different across computer brands and even
+     models. It is usually <kbd>F12</kbd>, but sometimes <kbd>Esc</kbd> or
+     another function key. If you are unsure, look at the POST screen that is
```
```suggestion
     models. It can be <kbd>F12</kbd>, but also <kbd>Esc</kbd>, <kbd>Enter</kbd> or
     another function key. If you are unsure, look at the POST screen that is
```

davidak replied:
thanks. i also added all keys you mentioned above

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+     :::
+
+     ::: {.note}
+     If your computer support both BIOS and UEFI boot, choose the UEFI option.
```
```suggestion
     If your computer supports both BIOS and UEFI boot, choose the UEFI option.
```

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+     and wait (or press <kbd>Enter</kbd> to speed up).
+
+4.   The graphical images will start their corresponding desktop environment
+     and the graphical installer, which can take some time. The minimal images
```
```suggestion
     and the graphical installer. The minimal images
```
It shouldn't really take long and be done in 30 seconds at most.

davidak replied:
When letting test people do the install, they where often impatient and thought they have to do something in the black screen before the desktop is shown... so i think this information is essential for people to patiently wait for the desktop to show up! (even if they might not read the manual)

Mic92 replied:
Yeah, kde sometimes take a bit to show something. Especially when you are not used to boot from USB.

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+     notebook, your model might be in the list. Select the language you are most
+     comfortable typing in.
+
+     ::: {.tip}
+     There are special keyboard layouts that are optimized to be ergonomic like
+     [NEO](https://www.neo-layout.org/),
+     [Dvorak](https://en.wikipedia.org/wiki/Dvorak_keyboard_layout) and
+     [Colemak](https://colemak.com/), in opposition to the classic QWERTY layout
+     which comes from typewriters and has not changed much since 1873.
+     You might want to check them out.
+     :::
+
```
```suggestion
```
I don't think we should promote that here. If a beginner chooses one of these and doesn't actually know what the consequences are then you cannot log in after a reboot.

davidak replied:
You are right. That information might not be useful and rather distracting here.

Should the mention of NEO in the existing documentation also be removed?

SuperSandro2000 replied:
We can probably keep it if it is something interesting other distros lack.

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+     You might want to check them out.
+     :::
+
+4.   On the "Users" screen, you have to type in your name, login name and
+     password. You can also enable an option to automatically login to the
```
```suggestion
4.   On the "Users" screen, you have to type a display name, login name and
     password. You can also enable an option to automatically login to the
```

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+     They are both popular choices and well tested on NixOS.
+     :::
+
+6.   You have the option to allow unfree software in the next screen.
```
Does this include unfree firmware? If not we should likely add a short section for that, too.

davidak replied:
yes. there is a detailed description in the installer

![Screenshot from 2022-05-30 02-51-00](https://user-images.githubusercontent.com/91113/170899056-3bd32229-4048-4ac1-b471-4ae340efe2c9.png)


(it does not enable unfree gpu drivers, but e.g. wifi drivers)

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+
+     ::: {.warning}
+     Make sure you have selected the correct disk at the top and that no valuable
+     data is still on the disk!
```
```suggestion
     data is still on the disk! It will be deleted when formatting the disk.
```

SuperSandro2000 commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,147 @@
+8.   Check the choices you made in the "Summary" and click "Install".
+
+     ::: {.note}
+     The install takes about 15 minutes. The time varies based on the selected
+     desktop, internet speed and disk write speed.
```
```suggestion
     The installation process can take about 15 minutes. The time varies based on the selected
     desktop environment, internet connection speed and disk read/write speed.
```

davidak replied:
it can take 8 minutes, it can take 10 minutes and it can take also 14 minutes, from my tests. but i have high-end hardware (nvme SSD) and gigabit internet, so i have choosen a higher number. if it's even faster, the user will be happy, but when it takes longer than this estimate, they might be unhappy. it says "takes **about** 15 minutes", so it's clear that it is not an exact value

i don't mentioned USB flash drive read speed and processor / memory speed, as it is not that relevant (install is mostly from internet and not much from usb drive i think). the section does not have to get so long

SuperSandro2000 commented on nixos/doc/manual/installation/installing-usb.section.md:
```diff
@@ -1,31 +1,54 @@
+
+::: {.warning}
+Etcher reports errors and usage statistics by default, which can be disabled in
+the settings. Some users might not like that.
```
```suggestion
the settings.
```

davidak replied:
I thought it might be relevant to note that there are people criticizing Etcher because of that. Some call it "spyware", even tho it does not collect personal data.

That's why i also list an alternative, which is sadly less user-friendly, because it has many versions to choose from (4 per arch).

i remove the sentence because it's quite opinionated

SuperSandro2000 commented on doc/build-aux/pandoc-filters/docbook-writer/html-elements.lua:
```diff
@@ -0,0 +1,11 @@
+Converts some HTML elements commonly used in Markdown to corresponding DocBook elements.
+]]
+
+function RawInline(elem)
+  if elem.format == 'html' and elem.text == '<kbd>' then
+    return pandoc.RawInline('docbook', '<keycap>')
+  elseif elem.format == 'html' and elem.text == '</kbd>' then
+    return pandoc.RawInline('docbook', '</keycap>')
+  end
+end
```
Thats handy! We might also extract that and merge this already.

davidak replied:
@jtojnar contributed it in https://github.com/davidak/nixpkgs/pull/1 :)

i hope merging this PR does not take too long

davidak commented:
@SuperSandro2000 thanks for your review! I addressed all comments.

dotlambda commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,142 @@
+     BIOS/UEFI settings to change the boot order. Again, search online for
+     details about your specific computer model.
+
+     For Apple computers with Intel processor press and hold the <kbd>⌥</kbd>
```
```suggestion
     For Apple computers with Intel processors press and hold the <kbd>⌥</kbd>
```

dotlambda commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,142 @@
+     the installed system.
+
+     ::: {.tip}
+     Leaving the language as "American English" will help to be able to copy
+     and paste error messages in a search engine or issue report.
```
```suggestion
     Leaving the language as "American English" will make it easier to search for
     error messages in a search engine or to report an issue.
```

dotlambda commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,142 @@
+     ::: {.tip}
+     If you don't have a favorite desktop and don't know which one to choose,
+     you can stick to either GNOME or Plasma. They have a quite different
+     design, so you should be able to choose which one you like better.
```
```suggestion
     design, so you should choose whichever you like better.
```

dotlambda commented on nixos/doc/manual/installation/installing.chapter.md:
```diff
@@ -1,30 +1,142 @@
+     selected desktop environment, internet connection speed and disk write speed.
+     :::
+
+9.  When the install is complete, reboot into your new system!
```
```suggestion
9.  When the install is complete, remove the USB flash drive and reboot into your new system!
```

davidak replied:
that works, thanks

i left it out because it should not not boot from usb again, but it is more elegant to close with this

davidak commented:
@dotlambda thanks for your review!

Mic92 commented on nixos/doc/manual/installation/installing-usb.section.md:
````diff
@@ -1,31 +1,54 @@
+
+```ShellSession
+sudo umount /dev/sdX*
+sudo dd if=<path-to-image> of=/dev/sdX bs=4M
````
That also ensures that all data is flushed before ejecting the drive:

```suggestion
sudo dd if=<path-to-image> of=/dev/sdX bs=4M conv=fsync
```

Mic92 replied:
This flag does not exists on macos though. So only add it for linux.

thiagokokada replied:
I generally use `oflag=sync` that seems to be similar, and I think it is available in macOS.

thiagokokada replied:
No, `oflag=sync` is a valid alternative for `fsync` (yeah, the effect is different but it still works as "ensure that all data is flushed"): https://unix.stackexchange.com/a/508726

What I meant is that `oflag` is not available in macOS. So changing to it wouldn't solve anything.

lourkeur replied:
FWIW Busybox's `dd` supports `conv=fsync` but not `oflag`.

Would a standalone `sync` command suffice?

samueldr replied:
I'm almost thinking at this point that it should be something like "add your preferred `dd` flags, if you're unsure, read the appropriate manual". As I stated, things are highly cargo-culted, and *always* causes this kind of discussion :). Though adding a final `sync` in that case would be beneficial to ensure that it's synced for people that won't look it up.

samueldr replied:
`conv=fdatasync` and `conv=sync` ends-up being functionally identical for the requirement of syncing.

As would calling `sync` after `dd` (AFAIUI), or calling `eject`.

Mic92 commented on nixos/doc/manual/installation/installing-usb.section.md:
````diff
@@ -1,31 +1,54 @@
+sudo dd if=<path-to-image> of=/dev/sdX bs=4M
+```
+
+Be careful to replace `sdX` with the correct device! You can use the `lsblk`
````
Users might read this to late :) Maybe lsblk should be moved up into the example?

Mic92 replied:
Maybe some example output and than what device was chosen:

```
$ lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
nvme0n1     259:0    0   1,8T  0 disk
├─nvme0n1p1 259:1    0  1000M  0 part /boot
├─nvme0n1p2 259:2    0    16M  0 part
├─nvme0n1p3 259:3    0 195,3G  0 part
└─nvme0n1p4 259:4    0   1,6T  0 part
```

davidak replied:
not sure how to move it up. having the output there would add a lot of clutter

should we just start with the task to find the right disk and explain the process in more detail?

on my system:

```
[davidak@gaming:~/code/nixpkgs]$ lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
sda           8:0    0 465.8G  0 disk 
├─sda1        8:1    0     8G  0 part [SWAP]
└─sda2        8:2    0 457.8G  0 part /nix/store
                                      /
sdb           8:16   1   3.7G  0 disk 
├─sdb1        8:17   1   2.2G  0 part 
└─sdb2        8:18   1     3M  0 part 
nvme0n1     259:0    0 931.5G  0 disk 
├─nvme0n1p1 259:1    0   512M  0 part 
├─nvme0n1p2 259:2    0 896.6G  0 part 
└─nvme0n1p3 259:3    0  34.4G  0 part
```

Mic92 replied:
Maybe this could be just:

```
# This command will print all disks, identify your usb stick device name from this list and use it in the commands below.
# WARNING: if you pick the wrong device you may override important data.
$ lsblk 
```

davidak replied:
i will add instructions to find the correct device as first step.

it's quiet unlikely that someone has `/dev/sdX` AND don't know what they are doing (just executing the command without understanding), but when someone just runs the command and get an error, our instructions are bad, so we should improve the experience

thanks for the suggestion!

davidak replied:
I have rewritten the complete sections in numbered list style.

davidak commented:
@Mic92 thanks for your review!

I have incorporated all feedback.

davidak commented:
Why is this still not merged?

Everyone is complaining that the documentation is outdated or incomplete and this PR would improve that, but now it's stalled for about 2 weeks. I had plans for further improvements once this is merged, but this contribution experience is very discouraging.

thiagokokada reviewed (commented):
A few nitpicks.

thiagokokada commented:
> Why is this still not merged?

> Everyone is complaining that the documentation is outdated or incomplete and this PR would improve that, but now it's stalled for about 2 weeks. I had plans for further improvements once this is merged, but this contribution experience is very discouraging.

Sorry, but commiters are also busy folks. And also we sometimes forget about PRs.

A reminder to review/merge a PR is always welcome, but please avoid an aggressive tone when doing so.

fufexan commented:
What's the status on this PR? It's been sitting for some time.

bobvanderlinden reviewed (approved):
This does seem to be an improvement over what is currently in the manual. The instructions for Linux seem to be correct (which is the part I can currently verify). Seems fine to have this available in the manual already.

davidak commented:
@bobvanderlinden thank you very much!

I rebased and built it again. Can we get another review and merge it? I don't want to miss another release.

thiagokokada commented:
@davidak Can you fix the two missing `.` that I found before? Otherwise LGTM.

davidak commented:
@thiagokokada i don't think a list of bullet points need a period, but i can do it for consistency. some have them when there are multiple sentences

thiagokokada commented:
> @thiagokokada i don't think a list of bullet points need a period, but i can do it for consistency. some have them when there are multiple sentences

Yeah, I was saying this mostly because only the first point on each was missing periods.

But I already fixed anyway, going to merge once I got back home.

davidak commented:
@thiagokokada sorry i did not get your point why that is useful.

I squashed your commit and added you to co-authors. I like that this is such a collaborative effort with 6 authors!

I also built it again, successfully.

thiagokokada commented:
> @thiagokokada sorry i did not get your point why that is useful.
> 
> I squashed your commit and added you to co-authors. I like that this is such a collaborative effort with 6 authors!
> 
> I also built it again, successfully.

I meant, the problem for me was the inconsistency, not correctness. It was only the first point on two different sections that were missing periods, while the other points in the same section included them.

You could either remove all of them or add the periods in the two places that I indicated, that was it.

thiagokokada commented:
Let's merge this before 22.11 release cut, so folks have time to improve on it until the release is finished.

thiagokokada commented:
Thanks @davidak and everyone else for the improvements.
